DeploymentClusteredTest.shouldDeployProcessAndCreateInstances

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? at least once
- Does it block your work? Blocked merging
- Do we suspect that it is a real failure? not sure

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
io.camunda.zeebe.it.clustering.DeploymentClusteredTest

[2021-10-13T09:22:45.475Z] [ERROR] io.camunda.zeebe.it.clustering.DeploymentClusteredTest.shouldDeployProcessAndCreateInstances  Time elapsed: 120.012 s  <<< ERROR!

[2021-10-13T09:22:45.475Z] org.junit.runners.model.TestTimedOutException: test timed out after 120 seconds

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/jdk.internal.misc.Unsafe.park(Native Method)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:234)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2123)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:133)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:108)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.FutureUtil.join(FutureUtil.java:21)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.future.CompletableActorFuture.join(CompletableActorFuture.java:196)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker.lambda$close$0(Broker.java:170)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker$$Lambda$1358/0x00000007c0977c40.accept(Unknown Source)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:753)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:731)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2108)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker.lambda$close$1(Broker.java:168)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker$$Lambda$1357/0x00000007c0977840.run(Unknown Source)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.LogUtil.doWithMDC(LogUtil.java:23)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker.close(Broker.java:163)
[2021-10-13T09:22:45.475Z] 	at app/zeebe.qa.integration.tests@1.3.0-SNAPSHOT/io.camunda.zeebe.it.clustering.ClusteringRule.lambda$after$3(ClusteringRule.java:257)
[2021-10-13T09:22:45.475Z] 	at app/zeebe.qa.integration.tests@1.3.0-SNAPSHOT/io.camunda.zeebe.it.clustering.ClusteringRule$$Lambda$1582/0x00000007c0a67840.accept(Unknown Source)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinPool$WorkQueue.helpCC(ForkJoinPool.java:1115)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinPool.externalHelpComplete(ForkJoinPool.java:1957)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurent.ForkJoinTask.tryExternalHelp(ForkJoinTask.java:378)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.externalAwaitDone(ForkJoinTask.java:323)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:412)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)

[2021-10-13T09:22:45.476Z] 	at app/zeebe.qa.integration.tests@1.3.0-SNAPSHOT/io.camunda.zeebe.it.clustering.ClusteringRule.after(ClusteringRule.java:254)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:59)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.FutureTask.run(FutureTask.java:264)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.lang.Thread.run(Thread.java:834)

[2021-10-13T09:22:45.476Z] 
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>
LOGS
</pre>
</details>


Failing https://ci.zeebe.camunda.cloud/blue/organizations/jenkins/camunda-cloud%2Fzeebe/detail/staging/1173/pipeline

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java
@@ -56,7 +56,6 @@ public final class ZeebePartition extends Actor
   private final PartitionAdminControl adminControl;
   private final PartitionTransition transition;
   private CompletableActorFuture<Void> closeFuture;
-  private ActorFuture<Void> currentTransitionFuture;
   private boolean closing = false;
 
   public ZeebePartition(
@@ -213,25 +212,20 @@ public final class ZeebePartition extends Actor
   }
 
   private void onRoleChange(final Role newRole, final long newTerm) {
-    ActorFuture<Void> nextTransitionFuture = null;
     switch (newRole) {
       case LEADER:
-        nextTransitionFuture = leaderTransition(newTerm);
+        leaderTransition(newTerm);
         break;
       case INACTIVE:
-        nextTransitionFuture = transitionToInactive();
+        transitionToInactive();
         break;
       case PASSIVE:
       case PROMOTABLE:
       case CANDIDATE:
       case FOLLOWER:
       default:
-        nextTransitionFuture = followerTransition(newTerm);
+        followerTransition(newTerm);
         break;
-    }
-
-    if (nextTransitionFuture != null) {
-      currentTransitionFuture = nextTransitionFuture;
     }
     LOG.debug("Partition role transitioning from {} to {} in term {}", raftRole, newRole, newTerm);
     raftRole = newRole;
@@ -288,9 +282,7 @@ public final class ZeebePartition extends Actor
 
   private ActorFuture<Void> transitionToInactive() {
     zeebePartitionHealth.setServicesInstalled(false);
-    final var inactiveTransitionFuture = transition.toInactive(context.getCurrentTerm());
-    currentTransitionFuture = inactiveTransitionFuture;
-    return inactiveTransitionFuture;
+    return transition.toInactive(context.getCurrentTerm());
   }
 
   private void registerListeners() {
@@ -463,7 +455,7 @@ public final class ZeebePartition extends Actor
     actor.run(
         () -> {
           if (!closing) {
-            currentTransitionFuture = transition.toInactive(context.getCurrentTerm());
+            transition.toInactive(context.getCurrentTerm());
           }
         });
   }
@@ -475,7 +467,7 @@ public final class ZeebePartition extends Actor
     actor.run(
         () -> {
           if (!closing) {
-            currentTransitionFuture = followerTransition(term);
+            followerTransition(term);
           }
         });
   }

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java
@@ -57,6 +57,7 @@ public final class ZeebePartition extends Actor
   private final PartitionTransition transition;
   private CompletableActorFuture<Void> closeFuture;
   private ActorFuture<Void> currentTransitionFuture;
+  private boolean closing = false;
 
   public ZeebePartition(
       final PartitionStartupAndTransitionContextImpl transitionContext,
@@ -143,38 +144,22 @@ public final class ZeebePartition extends Actor
 
   @Override
   protected void onActorClosing() {
-    context.getRaftPartition().getServer().removeSnapshotReplicationListener(this);
-    context.getRaftPartition().removeRoleChangeListener(this);
-
-    transitionToInactive()
+    // Already transitioned to inactive
+    startupProcess
+        .shutdown(actor, startupContext)
         .onComplete(
-            (nothing, err) -> {
-              context
-                  .getComponentHealthMonitor()
-                  .removeComponent(context.getRaftPartition().name());
+            (newStartupContext, error) -> {
+              if (error != null) {
+                LOG.error(error.getMessage(), error);
+              }
 
-              startupProcess
-                  .shutdown(actor, startupContext)
-                  .onComplete(
-                      (newStartupContext, error) -> {
-                        if (error != null) {
-                          LOG.error(error.getMessage(), error);
-                        }
+              // reset contexts to null to not have lingering references that could
+              // cause OOM problems in tests which start/stop partitions
+              startupContext = null;
+              context = null;
 
-                        // reset contexts to null to not have lingering references that could
-                        // cause OOM problems in tests which start/stop partitions
-                        startupContext = null;
-                        context = null;
-
-                        closeFuture.complete(null);
-                      });
+              closeFuture.complete(null);
             });
-  }
-
-  @Override
-  protected void onActorCloseRequested() {
-    LOG.debug("Closing ZeebePartition {}", context.getPartitionId());
-    context.getComponentHealthMonitor().removeComponent(zeebePartitionHealth.getName());
   }
 
   @Override
@@ -186,10 +171,19 @@ public final class ZeebePartition extends Actor
     closeFuture = new CompletableActorFuture<>();
 
     actor.run(
-        () ->
-            // allows to await current transition to avoid concurrent modifications and
-            // transitioning
-            currentTransitionFuture.onComplete((nothing, err) -> super.closeAsync()));
+        () -> {
+          LOG.debug("Closing ZeebePartition {}", context.getPartitionId());
+          closing = true;
+
+          removeListeners();
+          context.getComponentHealthMonitor().removeComponent(zeebePartitionHealth.getName());
+
+          final var inactiveTransitionFuture = transitionToInactive();
+
+          // allows to await current transition to avoid concurrent modifications and
+          // transitioning
+          inactiveTransitionFuture.onComplete((nothing, err) -> super.closeAsync());
+        });
 
     return closeFuture;
   }
@@ -210,7 +204,12 @@ public final class ZeebePartition extends Actor
   @Override
   @Deprecated // will be removed from public API of ZeebePartition
   public void onNewRole(final Role newRole, final long newTerm) {
-    actor.run(() -> onRoleChange(newRole, newTerm));
+    actor.run(
+        () -> {
+          if (!closing) {
+            onRoleChange(newRole, newTerm);
+          }
+        });
   }
 
   private void onRoleChange(final Role newRole, final long newTerm) {
@@ -298,6 +297,12 @@ public final class ZeebePartition extends Actor
     context.getRaftPartition().addRoleChangeListener(this);
     context.getComponentHealthMonitor().addFailureListener(this);
     context.getRaftPartition().getServer().addSnapshotReplicationListener(this);
+  }
+
+  private void removeListeners() {
+    context.getRaftPartition().removeRoleChangeListener(this);
+    context.getComponentHealthMonitor().removeFailureListener(this);
+    context.getRaftPartition().getServer().removeSnapshotReplicationListener(this);
   }
 
   @Override
@@ -455,14 +460,24 @@ public final class ZeebePartition extends Actor
     // restart from a new state. So we transition to Inactive to close existing services. The
     // services will be reinstalled when snapshot replication is completed.
     // We do not want to mark it as unhealthy, hence we don't reuse transitionToInactive()
-    actor.run(() -> currentTransitionFuture = transition.toInactive(context.getCurrentTerm()));
+    actor.run(
+        () -> {
+          if (!closing) {
+            currentTransitionFuture = transition.toInactive(context.getCurrentTerm());
+          }
+        });
   }
 
   @Override
   public void onSnapshotReplicationCompleted(final long term) {
     // Snapshot is received only by the followers. Hence we can safely assume that we have to
     // re-install follower services.
-    actor.run(() -> currentTransitionFuture = followerTransition(term));
+    actor.run(
+        () -> {
+          if (!closing) {
+            currentTransitionFuture = followerTransition(term);
+          }
+        });
   }
 
   public ActorFuture<Role> getCurrentRole() {

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/PartitionTransitionProcess.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/PartitionTransitionProcess.java
@@ -99,7 +99,11 @@ final class PartitionTransitionProcess {
   }
 
   ActorFuture<Void> prepare(final long newTerm, final Role newRole) {
-    LOG.info("Prepare transition from {} on term {} to {}", role, term, newRole);
+    LOG.info(
+        "Prepare transition from {} on term {} to {}",
+        context.getCurrentRole(),
+        context.getCurrentTerm(),
+        newRole);
     final ActorFuture<Void> prepareFuture = concurrencyControl.createFuture();
 
     if (stepsToPrepare.isEmpty()) {
@@ -119,8 +123,8 @@ final class PartitionTransitionProcess {
 
           LOG.info(
               "Prepare transition from {} on term {} to {} - preparing {}",
-              role,
-              term,
+              context.getCurrentRole(),
+              context.getCurrentTerm(),
               newRole,
               nextPrepareStep.getName());
 
@@ -143,7 +147,10 @@ final class PartitionTransitionProcess {
     }
 
     if (stepsToPrepare.isEmpty()) {
-      LOG.info("Preparing transition from {} on term {} completed", role, term);
+      LOG.info(
+          "Preparing transition from {} on term {} completed",
+          context.getCurrentRole(),
+          context.getCurrentTerm());
       future.complete(null);
 
       return;

--- a/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionTest.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionTest.java
@@ -36,6 +36,7 @@ import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
+import org.awaitility.Awaitility;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
@@ -266,6 +267,41 @@ public class ZeebePartitionTest {
     order.verify(transition).toLeader(0L);
     order.verify(transition).toInactive(anyLong());
     order.verify(raft).goInactive();
+  }
+
+  @Test
+  public void shouldCloseZeebePartitionWhileOngoingTransition() {
+    // given
+    final PartitionTransitionStep mockTransitionStep = mock(PartitionTransitionStep.class);
+    final CompletableActorFuture<Void> firstTransitionFuture = new CompletableActorFuture<>();
+    final CompletableActorFuture<Void> secondTransitionFuture = new CompletableActorFuture<>();
+    when(mockTransitionStep.transitionTo(any(), anyLong(), any(Role.class)))
+        .thenReturn(firstTransitionFuture)
+        .thenReturn(secondTransitionFuture)
+        .thenReturn(CompletableActorFuture.completed(null));
+
+    when(mockTransitionStep.prepareTransition(any(), anyLong(), any(Role.class)))
+        .thenReturn(CompletableActorFuture.completed(null));
+
+    transition =
+        spy(new PartitionTransitionImpl(List.of(mockTransitionStep, new NoopTransitionStep())));
+    partition = new ZeebePartition(ctx, transition, List.of(new NoopStartupStep()));
+    schedulerRule.submitActor(partition);
+
+    partition.onNewRole(Role.LEADER, 1);
+    schedulerRule.workUntilDone();
+
+    // when
+    final var closeFuture = partition.closeAsync();
+    schedulerRule.workUntilDone();
+    partition.onNewRole(Role.FOLLOWER, 2);
+    schedulerRule.workUntilDone();
+    firstTransitionFuture.complete(null);
+    secondTransitionFuture.complete(null);
+    schedulerRule.workUntilDone();
+
+    // then
+    Awaitility.await().until(closeFuture::isDone);
   }
 
   private static class NoopStartupStep implements PartitionStartupStep {

